docs/rfcs/xxxx-backwards-compatibility.md

```diff
@@ -0,0 +1,176 @@
+# Grafana Agent backwards compatibility 
+
+* Date: 2023-05-25
+* Author: Robert Fratto (@rfratto)
+* PR: [grafana/agent#3981](https://github.com/grafana/agent/pull/3981)
+* Status: Draft
+
+Grafana Agent has been following [semantic versioning](https://semver.org/)
+since its inception. After three years of development and 33 minor releases,
+the project is on trajectory to have a 1.0 release. 
+
+In the context of semantic versioning, a 1.0 release indicates that future
+minor releases have backwards compatibility with older minor releases in the
+same major release; version 1.1 is backwards compatible with version 1.0.
+Having major releases signals to users when they can upgrade with more
+confidence (minor releases) and when an upgrade may take more time (new major
+releases). 
+
+However, Grafana Agent is a large project, with a large surface area for what
+may be considered part of the backwards compatibility guarantees. This proposal
+formally establishes what parts of Grafana Agent will be protected by backwards
+compatibility. 
+
+## Goals 
+
+- Set expectations for what is covered by backwards compatibility. 
+- Set expectations for when upgrades to new major versions will be forced.
+
+## Proposal 
+
+Backwards compatibility means that a user can upgrade their version of Grafana
```

Comment:
If the Agent is extensible via modules and plugins, then backwards compatibility for the base Agent image would be much easier to achieve. That's more of a v2 or v3 thing, but maybe it is something you'd want to mention?

Reply:
That doesn't change the definition of backwards compatibility though: a new major version is required whenever there is a breaking change caused by updating Grafana Agent.

Reply:
True, although it does make it easier for both developers and customers:

- Developers won't have to create major versions as often
- Customers would be able to upgrade the core Agent binary without upgrading the version of plugins used in Agent configs. Moreover, they could mix and match plugins at different versions. Currently they have to use a set bundle of components, all at that version.

I know this might be a bit outside the scope of the RFC, but it could imply that component versions are one day treated separately and thought this might be worth mentioning in a "future" section for example. Up to you, I personally don't insist 😄